File: polynomial_regression.py
# Polynomial Regression

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset 
# x should be a matrix of features and y (dependent variable) should be a vector
# 1:2 represents 1st index, since upper bound is not considered
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Fitting linear regression to the dataset
from sklearn.linear_model import LinearRegression
linReg = LinearRegression()
linReg.fit(X, y)

# Fitting polynomial regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
polyReg = PolynomialFeatures(degree = 4)
X_poly = polyReg.fit_transform(X)
linReg2 = LinearRegression()
linReg2.fit(X_poly, y)

#Visualising the linear regression results
#Real observation points
plt.scatter(X, y, color = 'red')
plt.plot(X, linReg.predict(X), color = 'blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show() 

#Visualising the polynomial regression results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, linReg2.predict(polyReg.fit_transform(X_grid)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

#Predicting a new result with linear regression
linReg.predict(6.5)

#Predicting a new result with polynomial regression
linReg2.predict(polyReg.fit_transform(6.5)) 

File: polynomial_regression.R
# Polynomial Regression

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

#Fitting linear regression to the dataset
linReg = lm(formula = Salary ~ .,
            data = dataset)

#Fitting polynomial regression to the dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
polyReg = lm(formula = Salary ~ .,
             data = dataset)

#Visualizing the linear regression results
library(ggplot2)
ggplot() + 
    geom_point(aes(x = dataset$Level, y = dataset$Salary),
               color = 'red') +
    geom_line(aes(x = dataset$Level, y = predict(linReg, newdata = dataset)),
              color = 'blue') +
    ggtitle('Truth or bluff (Linear regression)') +
    xlab('Level') +
    ylab('Salary')

#Visualizing the polynomial regression results 
ggplot() + 
    geom_point(aes(x = dataset$Level, y = dataset$Salary),
               color = 'red') +
    geom_line(aes(x = dataset$Level, y = predict(polyReg, newdata = dataset)),
              color = 'blue') +
    ggtitle('Truth or bluff (Polynomial regression)') +
    xlab('Level') +
    ylab('Salary')

#Predicting a new result with linear regression 
y_pred = predict(linReg, data.frame(Level = 6.5))

#Predicting a new result with polynomial regression
y_pred = predict(polyReg, data.frame(Level = 6.5,
                                     Level2 = 6.5^2,
                                     Level3 = 6.5^3,
                                     Level4 = 6.5^4))
